From d0a670fba82ab6d90db25e050b3784ffa8f5e580 Mon Sep 17 00:00:00 2001
Message-Id: <d0a670fba82ab6d90db25e050b3784ffa8f5e580.1421886431.git.chang-joon.lee@intel.com>
In-Reply-To: <2a59efb195329ac28200fa6e68e1170a4cf8d03c.1421886431.git.chang-joon.lee@intel.com>
References: <2a59efb195329ac28200fa6e68e1170a4cf8d03c.1421886431.git.chang-joon.lee@intel.com>
From: Deepak S <deepak.s@intel.com>
Date: Fri, 12 Dec 2014 08:39:30 +0530
Subject: [PATCH 09/11] FOR_UPSTREAM [VPG]: drm/i915: Use vlv display irq
 setup code for chv

After Display power ungating, we need to restore display irq's.
Instead of creating new function for chv and duplicating the code.
Resuse valleyview_enable_display_irqs & valleyview_disable_display_irqs.

v2: Remove I915_LPE_PIPE_C_INTERRUPT bit from pipe_state (Thomas)

Issue: GMINL-4794
Change-Id: Ic974538b48ce69046185feebfdaad7cb5043c3ef
Signed-off-by: Deepak S <deepak.s@intel.com>
---
 drivers/gpu/drm/i915/i915_irq.c |   37 +++++++++++++++++++++++++++----------
 1 file changed, 27 insertions(+), 10 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_irq.c b/drivers/gpu/drm/i915/i915_irq.c
index 9fb02f2..9b70cad 100644
--- a/drivers/gpu/drm/i915/i915_irq.c
+++ b/drivers/gpu/drm/i915/i915_irq.c
@@ -4001,8 +4001,10 @@ static int ironlake_irq_postinstall(struct drm_device *dev)
 
 static void valleyview_display_irqs_install(struct drm_i915_private *dev_priv)
 {
+	struct drm_device *dev = dev_priv->dev;
 	u32 pipestat_mask;
 	u32 iir_mask;
+	int pipe;
 #ifdef CONFIG_SUPPORT_LPDMA_HDMI_AUDIO
 	u32 lpe_status_clear;
 #endif
@@ -4010,8 +4012,8 @@ static void valleyview_display_irqs_install(struct drm_i915_private *dev_priv)
 	pipestat_mask = PIPESTAT_INT_STATUS_MASK |
 			PIPE_FIFO_UNDERRUN_STATUS;
 
-	I915_WRITE(PIPESTAT(PIPE_A), pipestat_mask);
-	I915_WRITE(PIPESTAT(PIPE_B), pipestat_mask);
+	for_each_pipe(pipe)
+		I915_WRITE(PIPESTAT(pipe), pipestat_mask);
 	POSTING_READ(PIPESTAT(PIPE_A));
 
 	/* Enable the DPST interrupt also */
@@ -4021,16 +4023,22 @@ static void valleyview_display_irqs_install(struct drm_i915_private *dev_priv)
 			SPRITE1_FLIP_DONE_INT_STATUS_VLV |
 			PIPE_DPST_EVENT_STATUS;
 
-	i915_enable_pipestat(dev_priv, PIPE_A, pipestat_mask |
-					       PIPE_GMBUS_INTERRUPT_STATUS);
-	i915_enable_pipestat(dev_priv, PIPE_B, pipestat_mask);
+	i915_enable_pipestat(dev_priv, PIPE_A, PIPE_GMBUS_INTERRUPT_STATUS);
+	for_each_pipe(pipe)
+		i915_enable_pipestat(dev_priv, pipe, pipestat_mask);
 
 	iir_mask = I915_DISPLAY_PORT_INTERRUPT |
 			I915_DISPLAY_PIPE_A_EVENT_INTERRUPT |
 			I915_DISPLAY_PIPE_B_EVENT_INTERRUPT;
+
+	if (IS_CHERRYVIEW(dev))
+		iir_mask |= I915_DISPLAY_PIPE_C_EVENT_INTERRUPT;
 #ifdef CONFIG_SUPPORT_LPDMA_HDMI_AUDIO
 	iir_mask |=	(I915_LPE_PIPE_A_INTERRUPT |
 			I915_LPE_PIPE_B_INTERRUPT);
+
+	if (IS_CHERRYVIEW(dev))
+		iir_mask |= I915_LPE_PIPE_C_INTERRUPT;
 #endif
 	dev_priv->irq_mask &= ~iir_mask;
 
@@ -4043,6 +4051,9 @@ static void valleyview_display_irqs_install(struct drm_i915_private *dev_priv)
 			I915_HDMI_AUDIO_BUFFER_DONE;
 	I915_WRITE(I915_LPE_AUDIO_HDMI_STATUS_A, lpe_status_clear);
 	I915_WRITE(I915_LPE_AUDIO_HDMI_STATUS_B, lpe_status_clear);
+
+	if (IS_CHERRYVIEW(dev))
+		I915_WRITE(I915_LPE_AUDIO_HDMI_STATUS_C, lpe_status_clear);
 #endif
 
 	POSTING_READ(VLV_IER);
@@ -4050,13 +4061,18 @@ static void valleyview_display_irqs_install(struct drm_i915_private *dev_priv)
 
 static void valleyview_display_irqs_uninstall(struct drm_i915_private *dev_priv)
 {
+	struct drm_device *dev = dev_priv->dev;
 	u32 pipestat_mask;
 	u32 iir_mask;
+	int pipe;
 
 	iir_mask = I915_DISPLAY_PORT_INTERRUPT |
 		   I915_DISPLAY_PIPE_A_EVENT_INTERRUPT |
 		   I915_DISPLAY_PIPE_B_EVENT_INTERRUPT;
 
+	if (IS_CHERRYVIEW(dev))
+		iir_mask |= I915_DISPLAY_PIPE_C_EVENT_INTERRUPT;
+
 	dev_priv->irq_mask |= iir_mask;
 	I915_WRITE(VLV_IER, ~dev_priv->irq_mask);
 	I915_WRITE(VLV_IMR, dev_priv->irq_mask);
@@ -4068,14 +4084,15 @@ static void valleyview_display_irqs_uninstall(struct drm_i915_private *dev_priv)
 			PIPE_CRC_DONE_INTERRUPT_STATUS |
 			PIPE_DPST_EVENT_STATUS;
 
-	i915_disable_pipestat(dev_priv, PIPE_A, pipestat_mask |
-					        PIPE_GMBUS_INTERRUPT_STATUS);
-	i915_disable_pipestat(dev_priv, PIPE_B, pipestat_mask);
+	i915_disable_pipestat(dev_priv, PIPE_A, PIPE_GMBUS_INTERRUPT_STATUS);
+	for_each_pipe(pipe)
+		i915_disable_pipestat(dev_priv, pipe, pipestat_mask);
 
 	pipestat_mask = PIPESTAT_INT_STATUS_MASK |
 			PIPE_FIFO_UNDERRUN_STATUS;
-	I915_WRITE(PIPESTAT(PIPE_A), pipestat_mask);
-	I915_WRITE(PIPESTAT(PIPE_B), pipestat_mask);
+
+	for_each_pipe(pipe)
+		I915_WRITE(PIPESTAT(pipe), pipestat_mask);
 	POSTING_READ(PIPESTAT(PIPE_A));
 }
 
-- 
1.7.9.5

